EIPS/eip-7761.md

```diff
@@ -0,0 +1,107 @@
+
+If `target_address` points to an account with a contract mid-creation, it behaves aligned with similar instructions like `EXTCODESIZE` and returns `0`.
+
+## Rationale
+
+### Alternative solutions
+
+There have been other solutions proposed to alleviate the problems related to lack of code introspection required for ERC-721 and ERC-1155 standards:
+
+1. Extra status code for `EXT*CALL` instruction - allowing to discriminate a result coming from calling an EOA
+2. Extra argument for `EXT*CALL` (a "fail if EOA" flag)
+3. Two return values from `EXT*CALL` (status code + whether it was EOA)
+4. `EXT*CALL` setting a new `callstatus` register (+ a new `CALLSTATUS` instruction)
+5. Reenable `EXTCODESIZE` in EOF, keeping its behavior same as in legacy
+
+`ISCONTRACT` has been chosen as the most elegant and minimal solution satisfying the requirements at hand and still able to be introduced in EOFv1.
+
+### Reuse the `0x3b` (`EXTCODESIZE`) opcode for `ISCONTRACT`
+
+A new opcode is prefered by a general policy to not reuse opcodes. Also `ISCONTRACT` can be rolled out in legacy EVM if desired.
+
+### Keep code introspection banned
+
+Removing code introspection is one of the tenets of EOF and `ISCONTRACT` would be an exception from the principle. Without `ISCONTRACT`, ERC-721 and ERC-1155 standard implementations have to resort to either:
+
+1. Leveraging a "booster contract" which would be legacy and would call `EXTCODESIZE` for them. This has been deemed inelegant and inconvenient from the point of view of library implementers, requiring them to hard code an address of such a contract (with the usual address-related problems arising on different EVM chains)
+2. Continuing to use legacy EVM themselves. This is sub-optimal, since EVM compilers are likely to at some point deprecate legacy EVM as compilation target
+3. Updating the standards to not rely on code introspection (`safeTransfer` safe guards) patterns, also unlikely to happen as ERC-721 and ERC-1155 are Final and adopted in practice.
+
+### "Endgame Account Abstraction" issues
+
+TBD - can someone verbalize it?
+    
+### Relation to [EIP-7702](./eip-7702.md) "Set EOA account code"
+    
+After [EIP-7702](./eip-7702.md) is activated, the discrimination between EOAs and contract accounts using `EXTCODESIZE` (or `ISCONTRACT`) has an edge case: Whenever an EOA sets its code to a contract account which does not respond as expected to an `onERC721Received` (`onERC1155Received`) callback, transfers to it will revert, despite the recipient being able to interact with the token. This has been deemed unlikely to be a problem, as for the intended real-world uses of `EIP-7702`, those callbacks will be implemented by designator codes.
```

Comment:
```suggestion
After [EIP-7702](./eip-7702.md) is activated, the discrimination between EOAs and contract accounts using `EXTCODESIZE` (or `ISCONTRACT`) has an edge case: Whenever an EOA sets its code to a contract account which does not respond as expected to an `onERC721Received` (`onERC1155Received`) callback, transfers to it will revert, despite the recipient being able to interact with the token. This has been deemed unlikely to be a problem, as for the intended real-world uses of EIP-7702, those callbacks will be implemented by designator codes.
```